### DISCUSSION TITLE
Element indexing issue in MaterialBase system

### DISCUSSION THREAD

**Discussion by feltner515** [idaholab/moose#25059] [Q&A General] [answered]
Hi Moose users

I have been working on implementing block-restricted advanced exodus restart. 

https://github.com/idaholab/moose/discussions/24760

As part of this, I have been editing `initQpStatefulProperties` for several material models to allow me to specify an initial state for all quadrature points from an old simulation. To do this, I am using SolutionUserObjects and SolutionFunctions to create GenericFunctionMaterials to give me an initial solution at all Qp's. The edits to the code are very simple, for example:
`_hardening_var_init(this->template getGenericMaterialProperty<Real, is_ad>("hardening_var_init1"))`
to get the generic material property to initialize a hardening variable, and :
```
_hardening_variable[_qp] = 0;
_hardening_variable[_qp] += _hardening_var_init[_qp];
```
 in `initQpStatefulProperties` to initialize the values.

This exercise has exposed an issue with the element indexing when 'initQpStatefulProperties' is called in the material base. There seems to be a one element difference between the "current element" that is being operated on, and the "current element" that the initial value is being read from. To show this I created a material from a parsed function where **the value is equal to the x coordinate of the Qp**. The GenericFunctionMaterial looks like this, as you'd expect:
![image](https://github.com/idaholab/moose/assets/85691462/19b34c9a-9cac-4233-aaef-552ae23a5e30)
The actual initial state of the variable looks like this:
![image](https://github.com/idaholab/moose/assets/85691462/5db61558-af1e-4a5f-a2b0-8b64b00e1f3a)
I may be wrong, but I do not think any of my additions are causing this. I planned on getting to the root cause of this, but I wanted to post this here so someone more qualified than me can see about getting a fix for this committed. The implications are only important if you are trying to use `initQpStatefulProperties` for something other than a constant field.

Thank you
Langdon

**Comment by GiudGiud** [comment]
@loganharbour

**Comment by loganharbour** [comment]
> To do this, I am using SolutionUserObjects and SolutionFunctions to create GenericFunctionMaterials to give me an initial solution at all Qp's. The edits to the code are very simple, for example:
_hardening_var_init(this->template getGenericMaterialProperty<Real, is_ad>("hardening_var_init1"))

This confuses me... that method is const. Are you const casting somewhere? Are you initializing a material property in a place in which the material wasn't declared?

I follow everything else you're doing and might have a few other concerns... but need to know the above first.

**Reply by feltner515** [reply]
Hi Logan, thank you for following up with me!

I am casting this as a const in the header file, 

`const GenericMaterialProperty<Real, is_ad> & _hardening_var_init;`  

I am not explicitly declaring the material property, rather just getting a material property named "hardening_var_init1". This is kind of a hacky solution and I do plan to clean this up later on once I get the method working.

Everything seems to work fine up until the point where I am calling hardening_var_init1 into `initQpStatefulProperties`. It seems like it is putting each cooresponding solution on the element one index higher than it is supposed to be.

**Reply by feltner515** [reply]
I think I have figured out the issue. I changed 
`_hardening_variable[_qp] += _hardening_var_init[_qp];`
to 
`_hardening_variable[_qp] += _current_elem -> id();`
yielding this:
![image](https://github.com/idaholab/moose/assets/85691462/77e55511-1c0b-440e-93e9-ee96455522be)
The actual element id's go from 1 to 100, while `initQpStatefulProperties` is operating on elements 0 to 99, causing the one element offset. Is this mismatch to be expected, and does element numbering necessarily need to start at 0. The mesh in this model is created with a GeneratedMeshGenerator.

**Reply by loganharbour** [reply]
> The actual element id's go from 1 to 100,

This isn't true. element IDs start with 0. Exodus output is 1 indexed, so they show up starting + 1 in exodus.

**Reply by feltner515** [reply, marked as answer]
In the meantime I figured this out. I'll put the solution here to bookend this thread in case anyone else runs into the same issue in the future. 

Basically my material I was bringing in was indexed in exodus id's, so it would show up correctly on its own in exodus outputs but was one index off in the kernel. To solve this, instead of creating a generic function material and calling the material into the kernel, call the function itself into the kernel. 

`_functions00(FunctionInterface::getFunctionByName(_base_name + "solution_fcn_hardening_var")),`

Then you can evaluate the function at each Qp within the kernel and assign the value so there is never a chance for the indexing inconsistency to pop up.

`_hardening_variable[_qp]=_functions.value(0,_q_point[_qp]);`

I have block-restricted stateful property initialization from exodus working now using this basic pattern.

Thank you for your help Logan!
